src/main/java/eu/dissco/core/digitalspecimenprocessor/service/ProcessingService.java

```diff
@@ -399,29 +423,62 @@ private void rollbackNewSpecimen(DigitalSpecimenRecord digitalSpecimenRecord,
       }
     }
     repository.rollbackSpecimen(digitalSpecimenRecord.id());
-    fdoRecordBuilder.rollbackHandleCreation(digitalSpecimenRecord);
     try {
       kafkaService.deadLetterEvent(
           new DigitalSpecimenEvent(enrichments, digitalSpecimenRecord.digitalSpecimen()));
     } catch (JsonProcessingException e) {
-      log.error("Fatal exception, unable to dead letter queue: " + digitalSpecimenRecord.id(), e);
+      log.error(DLQ_FAILED + digitalSpecimenRecord.id(), e);
     }
   }
 
-  private DigitalSpecimenRecord mapToDigitalSpecimenRecord(DigitalSpecimenEvent event) {
+  private void rollbackHandleCreation(List<DigitalSpecimenRecord> records){
+    var request = fdoRecordBuilder.buildRollbackCreationRequest(records);
     try {
-      return new DigitalSpecimenRecord(
-          fdoRecordBuilder.createNewHandle(event.digitalSpecimen()),
-          midsService.calculateMids(event.digitalSpecimen()),
-          1,
-          Instant.now(),
-          event.digitalSpecimen()
-      );
-    } catch (TransformerException | PidCreationException e) {
-      log.error("Failed to process record with id: {}",
-          event.digitalSpecimen().physicalSpecimenId(),
-          e);
+      handleComponent.rollbackHandleCreation(request);
+    } catch (PidCreationException | PidAuthenticationException e) {
+      var ids = records.stream().map(DigitalSpecimenRecord::id).toList();
+      log.error("Unable to rollback handles for new specimens. Bad handles: {}", ids);
+    }
+  }
+
+  private void filterUpdatesAndRollbackHandles(List<UpdatedDigitalSpecimenTuple> records){
+    var recordsToRollback = records.stream()
+        .filter(r -> fdoRecordBuilder.handleNeedsUpdate(r.currentSpecimen().digitalSpecimen(),
```

Comment:
Looks like you for both taking the currentSpecimen. I think that is incorrect as it will never indicate it should update. One needs to be new, the other the current (indicates it was updated) then we update back to the currentSpecimen (where we at first updated to the new record)

Reply:
Good catch!